### PR TITLE
add modbus thing

### LINKED ISSUE #7: Modbus Test Things
We should Modbus Test Things since it is one of the bindings at https://w3c.github.io/wot-binding-templates/bindings/protocols/modbus. However, it cannot be based on the calculator since it is more property-driven. This is gonna be the case for BACnet so we can use the same TM.

Also see https://github.com/yaacov/node-modbus-serial#modbustcp-server

### PR DESCRIPTION
**Pull request by hasanheroglu**
Closes #7. 

- Needed to change docker image for modbus, since alpine image failed to install @serialport/bindings binary.
- New modbus thing publishes its TD using express on port  "3" + <port_number_of_the_thing>.

### CODE REVIEW

File: things/lift/modbus/js/main.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        if (thingUnitID === unitID) {
+            // trying to change floor number
+            if (addr === 1) {
+                // lift is on the move
+                if (discreteInputs[0]) {
+                    console.log("Lift is on the move, cannot change the floor number")
+                } else {
+                    console.log(`Changing the floor number to ${value}`)
+                    holdingRegisters[addr - 1] = value
+                    discreteInputs[0] = 1
```
Comment (egekorkan) [inline things/lift/modbus/js/main.js:142]:
```suggestion
                    // simulating movement
                    discreteInputs[0] = 1
```

File: things/lift/modbus/js/main.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+thingDescription['properties']['floorNumber']['forms'] = floorNumberForms
+
+const vector = {
+    getDiscreteInput: function(addr, unitID) {
+        if (thingUnitID === unitID) {
+            console.log(`Reading discrete input @${addr}`)
+            return discreteInputs[addr - 1];
+        }
+    },
+    getInputRegister: function(addr, unitID) {
```
Comment (egekorkan) [inline things/lift/modbus/js/main.js:113]: can be removed since it is not used

**Review by egekorkan — CHANGES_REQUESTED** [review]
- [ ] Changing lift to elevator
- [ ] TD should saved to a file with localhost hostname

I will test it locally and comment further

File: things/lift/modbus/js/main.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+discreteInputs[0] = 0
+thingDescription['properties']['onTheMove']['forms'] = onTheMoveForms
+
+const floorNumberForms = [{
+    "href": `${thingUnitID}?address=1&quantity=1`,
+    "op": [
+        "readproperty"
+    ],
+    "modbus:entity": "HoldingRegister",
+    "modbus:function": "readHoldingRegisters",
```
Comment (egekorkan) [inline things/lift/modbus/js/main.js:92]:
```suggestion
    "modbus:function": "readHoldingRegister",
```

File: things/lift/modbus/js/main.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+const thingDescription = placeholderReplacer.replace(thingModel)
+thingDescription['@type'] = 'Thing'
+
+const coils = new Array(9999)
+const discreteInputs = new Array(9999)
+const inputRegisters = new Array(9999)
+const holdingRegisters = new Array(9999)
+
+const lightSwitchForms = [{
+    "href": `${thingUnitID}?address=1&quantity=1`,
```
Comment (egekorkan) [inline things/lift/modbus/js/main.js:53]: unitID should be at the base

File: things/lift/modbus/js/main.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    "op": [
```
Comment (egekorkan) [inline things/lift/modbus/js/main.js:54]: no need for array (for all forms) if there is only one value

File: things/lift/modbus/js/main.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                    setTimeout(() => {
+                        discreteInputs[0] = 0
+                    }, 5000)
+                }
+            }
+        }
+
+        return
+    },
+    setCoil: function(addr, value, unitID) {
```
Comment (egekorkan) [inline things/lift/modbus/js/main.js:152]: a log of the change would be nice

**Review by egekorkan — CHANGES_REQUESTED** [review]
for others, https://github.com/favalex/modbus-cli can be used for testing at the moment

Together with @hasanheroglu we did a test where I have opened a port on my local network and he sent the requests over the internet, using the CLI above. This way, we can really have a simulated Modbus device exposed to the internet

File: things/lift/lift.tm.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+         "observable": true
+      },
+      "onTheMove": {
+         "type": "boolean",
+         "readOnly": true,
+         "writeOnly": false,
+         "observable": true
+      },
+      "floorNumber": {
+         "type": "integer",
```
Comment (egekorkan) [inline things/lift/lift.tm.json:35]: we need minimum maximum for this

Comment (hasanheroglu) [reply things/lift/lift.tm.json]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/lift/modbus/js/main.js (hunk 6 above)
Comment (hasanheroglu) [reply things/lift/modbus/js/main.js]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/lift/modbus/js/main.js (hunk 5 above)
Comment (hasanheroglu) [reply things/lift/modbus/js/main.js]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/lift/modbus/js/main.js (hunk 4 above)
Comment (hasanheroglu) [reply things/lift/modbus/js/main.js]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/lift/modbus/js/main.js (hunk 3 above)
Comment (hasanheroglu) [reply things/lift/modbus/js/main.js]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/lift/modbus/js/main.js (hunk 2 above)
Comment (hasanheroglu) [reply things/lift/modbus/js/main.js]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/lift/modbus/js/main.js (hunk 1 above)
Comment (hasanheroglu) [reply things/lift/modbus/js/main.js]: Fixed at [1e7bc99](https://github.com/eclipse-thingweb/test-things/pull/9/commits/1e7bc99a0e815868f9254edfff03ef9562635cd2)

File: things/elevator/modbus/js/main.js [hunk 8]
```diff
@@ -70,76 +63,104 @@ const lightSwitchForms = [{
...
-        "writeproperty"
-    ],
+    "href": `?address=1&quantity=1`,
+    "op": "writeproperty",
     "modbus:entity": "HoldingRegister",
     "modbus:function": "writeSingleHoldingRegister",
     "contentType": "application/octet-stream"
 }]
 
+const minFloorNumber = 0
```
Comment (egekorkan) [inline things/elevator/modbus/js/main.js:98]: These should be in the TM and TD as well

**Review by egekorkan — CHANGES_REQUESTED** [review]
Also, the TD should be submitted with this PR

File: things/elevator/modbus/js/main.js (hunk 8 above)
Comment (hasanheroglu) [reply things/elevator/modbus/js/main.js:98]: There should be `minimum` and `maximum` fields for the property floorNumber with new commits.

**Review by egekorkan — APPROVED** [review]

**Comment by egekorkan** [comment]
@relu91 Can you provide your opinion too? It is written after the latest binding template and not conform to node-wot actually but a next step would be to update node-wot as well

**Comment by egekorkan** [comment]
I will double check the conflict and merge it to try hosting

**Pull request merged**